fms/models/gpt_oss.py

```diff
@@ -0,0 +1,652 @@
+                m.reset_parameters()
+
+    def _clean_up_rot_emb_cache(
+        self,
+        cached_freqs: dict[Optional[torch.device], dict[int, torch.Tensor]],
+        max_seq_len_cached: dict[Optional[torch.device], int],
+    ):
+        # remove meta tensors from cached_freqs
+        for dev in list(cached_freqs.keys()):
+            for alp in list(cached_freqs[dev].keys()):
+                if cached_freqs[dev][alp][0].device == torch.device("meta"):
+                    del cached_freqs[dev][alp][0]  # type: ignore
+                    if len(cached_freqs[dev]) == 0:
+                        del cached_freqs[dev]
+                        del max_seq_len_cached[dev]
+
+    def post_init(self):
+        # This function is called in `get_model` after the model is
+        # fully initalized on the correct device
+
+        self._clean_up_rot_emb_cache(
+            self.rot_emb.cached_freqs,
+            self.rot_emb.max_seq_len_cached,
+        )
+
+        # RoPE init
+        for device in set(
+            [param.device for param in self.parameters()]
+            + [buffer.device for buffer in self.buffers()]
+        ):
+            with torch.no_grad():
+                self.rot_emb.compute_freqs_cis(
+                    device, 2 * self.config.max_expected_seq_len
+                )
+
+    def forward(
+        self,
+        x_in: torch.Tensor,
+        position_ids: Optional[torch.Tensor] = None,
+        past_key_value_states: Optional[
+            Tuple[torch.Tensor,] | list[torch.Tensor] | list[None]
+        ] = None,
+        use_cache: bool = False,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len x emb_dim if input is already embedded, otherwise batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+
+        attn_kwargs["attn_name"] = attn_kwargs.get("attn_name", "sdpa_with_sinks")
+
+        get_attention_type(**attn_kwargs)["validate_attn_kwargs"](
+            input_ids=x_in,
+            position_ids=position_ids,
+            past_key_value_states=past_key_value_states,
+            **attn_kwargs,
+        )
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        if x_in.dim() == 2:  # input is not already embedded
+            x_in = self.embedding(x_in)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            output = layer(
+                x=x_in,
+                position_ids=position_ids,
+                past_key_value_state=past_key_value_states[i],
+                use_cache=use_cache,
+                sliding_window=layer.layer_sliding_window,
```

Comment:
more of a design question, but why would we call the layer forward and pass in layer.layer_sliding_window? The layer itself should have that information already, you can just use it.